### PR TITLE
API for generate, cancel ewb and settings doctype created

### PR DESCRIPTION
**Pull request by KaviyaPeriyasamy**
Screenshot from 2020-10-06 10-32-27
Screenshot from 2020-10-06 10-32-18

### CODE REVIEW

File: ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) 2020, Aerele and contributors
+// For license information, please see license.txt
+
+frappe.ui.form.on('EWB API Integration Settings', {
+	onload: function(frm) {
+		frm.set_df_property("gsp","options",['','Adaequare'].join('\n'))
```
Comment (karthikeyan5) [inline ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.js]: add this in the doctype json

File: ewb_api_integration/ewb_api_integration/gsp/adaequare.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+from frappe import _
+from frappe.utils import cint
+from ewb_api_integration.ewb_api_integration.doctype.ewb_api_integration_settings.ewb_api_integration_settings import get_ewb_api_config
+
+base_url = 'https://gsp.adaequare.com'
+
+def generate_ewb(config_doc, ewb, dt, dn):
+	data = make_supporting_request_data(ewb['billLists'][0])
+	data.update(calculate_amounts(dt, dn))
+	generate_url = '/test/enriched/ewb/ewayapi?action=GENEWAYBILL'
```
Comment (karthikeyan5) [inline ewb_api_integration/ewb_api_integration/gsp/adaequare.py]: keep as dict in one place

File: ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.json [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+   "label": "Select GSP"
+  },
+  {
+   "depends_on": "eval:doc.gsp=='Adaequare'",
+   "fieldname": "username",
+   "fieldtype": "Data",
+   "label": "Username"
+  },
+  {
+   "depends_on": "eval:doc.gsp=='Adaequare'",
```
Comment (karthikeyan5) [inline ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.json]: remove depends on

**Review by karthikeyan5 — CHANGES_REQUESTED** [review]
- add multiple GSTIN support

File: ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.json (hunk 3 above)
Comment (KaviyaPeriyasamy) [reply ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.json]: Requested changes updated in the below commits. Can you please review again?

File: ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+from frappe.model.document import Document
+import barcode
+from frappe import _
+from frappe.custom.doctype.custom_field.custom_field import create_custom_fields
+from erpnext.regional.india.utils import get_gst_accounts, get_itemised_tax_breakup_data
+
+class EWBAPIIntegrationSettings(Document):
+	pass
+
+def set_field_values(doc, action):
```
Comment (karthikeyan5) [inline ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.py:17]: do this on validate, after checking docstatus

File: ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			"allow_on_submit": 1,
+			"read_only": 1,
+			"hidden": 1
+			}
+		]
+	}
+	create_custom_fields(custom_fields, ignore_validate=frappe.flags.in_patch, update=True)
+
+# Calculation might be client specific. Will not work if the item specific taxes included.
+def calculate_amounts(dt, dn):
```
Comment (karthikeyan5) [inline ewb_api_integration/ewb_api_integration/doctype/ewb_api_integration_settings/ewb_api_integration_settings.py:46]: check the jinja sent to you and change accodingly.

File: ewb_api_integration/ewb_api_integration/doctype/gstin_credential_mapping/gstin_credential_mapping.json [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+   "in_list_view": 1,
+   "label": "GSTIN",
+   "unique": 1
+  },
+  {
+   "fieldname": "username",
+   "fieldtype": "Data",
+   "in_list_view": 1,
+   "label": "Username",
+   "unique": 1
```
Comment (karthikeyan5) [inline ewb_api_integration/ewb_api_integration/doctype/gstin_credential_mapping/gstin_credential_mapping.json]: remove unique for username and password

File: ewb_api_integration/ewb_api_integration/gsp/adaequare.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+			'staging_generate_url': '/test/enriched/ewb/ewayapi?action=GENEWAYBILL',
+			'live_generate_url': '/enriched/ewb/ewayapi?action=GENEWAYBILL',
+			'staging_cancel_url': '/test/enriched/ewb/ewayapi?action=CANEWB',
+			'live_cancel_url': '/enriched/ewb/ewayapi?action=CANEWB',
+			'staging_update_transporter_url': '/test/enriched/ewb/ewayapi?action=UPDATETRANSPORTER',
+			'live_update_transporter_url': '/enriched/ewb/ewayapi?action=UPDATETRANSPORTER',
+			'staging_get_ewb_url': '/test/enriched/ewb/ewayapi/GetEwayBill',
+			'live_get_ewb_url': '/enriched/ewb/ewayapi/GetEwayBill'}
+
+def generate_ewb(ewb, dt, dn):
```
Comment (karthikeyan5) [inline ewb_api_integration/ewb_api_integration/gsp/adaequare.py]: remove dt,dn dependency.

similarly for all other functions.

File: ewb_api_integration/hooks.py [hunk 8]
```diff
@@ -143,3 +143,17 @@
...
 # auto_cancel_exempted_doctypes = ["Auto Repeat"]
 
+doctype_js = {
+    "Sales Invoice" : "public/js/sales_invoice.js",
+}
+
+doc_events = {
+	"Sales Invoice": {
+		"on_submit": "ewb_api_integration.ewb_api_integration.doctype.ewb_api_integration_settings.ewb_api_integration_settings.set_field_values",
+		"on_update_after_submit": "ewb_api_integration.ewb_api_integration.gsp.gsp.update_transporter",
```
Comment (karthikeyan5) [inline ewb_api_integration/hooks.py:153]: error should prevent update.

**Review by karthikeyan5 — CHANGES_REQUESTED** [review]

File: ewb_api_integration/hooks.py (hunk 8 above)
Comment (KaviyaPeriyasamy) [reply ewb_api_integration/hooks.py:153]: ```on_update_after_submit``` itself prevents the update incase of any error.

Comment (KaviyaPeriyasamy) [reply ewb_api_integration/hooks.py:153]: Can you review the below commits? It will resolve the rem above.

**Review by karthikeyan5 — APPROVED** [review]
note: some changes to be made after merging this. check next commit after merge.

**Pull request merged**